torch/_dynamo/compiled_autograd.py

```diff
@@ -98,7 +104,7 @@ def tensor_pre_hook(self, inputs, hook_id, i: int):
             inputs[i],
         )
         with disable_proxy_modes_tracing():
-            inputs[i] = inputs[i].clone()
+            inputs[i] = maybe_clone(inputs[i])
```

Comment:
It is not actually putting a clone in the graph due to `disable_proxy_modes_tracing`.  The result of the clone is the fake tensor return value of the hook.  We don't call the hook, so we need to create a new fake tensor to bind the hook result to.

Reply:
I'm not sure how do we handle the case where the user hook returns `None` meaning that we should keep the input as-is?
Do we want to wrap the hook function added to the graph with code handling that as Dynamo should be aware of that check?

Also always doing a clone here might not match the view/layout semantic from the real backward. Shouldn't we run the real hook here to get a good output?

Reply:
Returning None is handled by:
https://github.com/pytorch/pytorch/blob/71d18f61059dbb617a8afc79b64598d2fbb89d63/torch/_dynamo/external_utils.py#L22-L29

It would be unsafe to call the real hook here, because it might execute arbitrary Python code that would have side effects.  Instead, we put a call to `call_hook` in the graph and let dynamo handle the Python stuff.

The size/view/layout may be wrong here if the hook does something to change them.  However, this usually won't matter because after dynamo we retrace everything again with AotAutograd, which will recalculate all the view/stride data after the hooks are inlined.

We have had some discussions about this in the past.  Not sure if there is a better solution.